discord.go

```diff
@@ -0,0 +1,266 @@
+	// rest of this function.
+	if s := ch.SilentTime(); event.Timestamp.Before(s) {
+		log.DebugContext(ctx, "channel is silent", slog.Time("until", s))
+		return
+	}
+
+	// TODO: Learn from this user only if not disabled, not sure if applicable to discord
+	// TODO: Meme detector stuff, possibly not necessary in discord?
+
+	// Empty messages could exist due to being a sticker or a file upload
+	if event.Content != "" {
+		robo.discordLearn(ctx, ch.Learn, event)
+	}
+
+	// Now we can check rate limits
+	t := time.Now()
+	r := ch.Rate.ReserveN(t, 1)
+	if d := r.DelayFrom(t); d > 0 {
+		log.InfoContext(ctx, "rate limited",
+			slog.String("action", "speak"),
+			slog.String("delay", d.String()),
+		)
+		r.CancelAt(t)
+		return
+	}
+
+	// Attempt to handle textual commands
+	if strings.HasPrefix(event.Content, session.State.User.Mention()) {
+		// Message starts by mentioning robot
+		trimmed := strings.TrimSpace(strings.TrimPrefix(event.Content, session.State.User.Mention()))
+		parser := regexp.MustCompile(`^(?i:say|generate)\s*(?i:something)?\s*(?i:starting)?\s*(?i:with)?\s+(.*)`)
+		matches := parser.FindStringSubmatch(trimmed)
+		if matches != nil && len(matches) > 1 {
+			prompt := matches[1]
+			robo.discordSend(ctx, session, event.Message.Reference(), event.ChannelID, ch.Send, prompt)
+		} else {
+			robo.discordSend(ctx, session, event.Message.Reference(), event.ChannelID, ch.Send, "")
+		}
+		return
+	}
+
+	// Not rate limited and not a command so check for random message chance
+	if rand.Float64() > ch.Responses {
+		return
+	}
+	robo.discordSend(ctx, session, nil, event.ChannelID, ch.Send, "")
+}
+
+func (robo *Robot) discordSend(ctx context.Context, session *discordgo.Session, reference *discordgo.MessageReference, channelId string, tag string, prompt string) {
+	response := robo.discordThink(ctx, tag, prompt)
+	if response == "" {
+		return
+	}
+	message := &discordgo.MessageSend{
+		Content:         response,
+		AllowedMentions: nil,
+		Reference:       reference,
+	}
+	_, err := session.ChannelMessageSendComplex(
+		channelId,
+		message,
+	)
+	if err != nil {
+		_ = fmt.Errorf("failed to send discord message: %w", err)
+	}
+}
+
+func (robo *Robot) discordThink(ctx context.Context, tag string, prompt string) string {
+	m, _, err := brain.Think(ctx, robo.brain, tag, prompt)
+	if err != nil {
+		_ = fmt.Errorf("couldn't think: %w", err)
+		return ""
+	}
+	return m
+}
+
+func (robo *Robot) discordLearn(ctx context.Context, tag string, event *discordgo.MessageCreate) {
+	userHash := robo.hashes().Hash(event.Author.ID, event.GuildID, event.Timestamp)
+	message := brain.Message{
+		Sender:      userHash,
+		ID:          event.Message.ID,
+		To:          event.GuildID,
+		Text:        event.Content,
+		Timestamp:   event.Timestamp.UnixMilli(),
+		IsModerator: event.Member.Permissions&discordgo.PermissionManageMessages != 0,
+		IsElevated:  false, // TODO: Possibly some other flag?
```

Comment:
Eventually this would be a check for some role configured for the server, but that's out of scope for this PR. (It's still out of scope for Twitch, even; nothing reads this field.)